answers.go

```diff
@@ -281,10 +281,22 @@ func (answers *Answers) MatchingExact(qtype uint16, clientIp string, fqdn string
 	}
 }
 
+// Shuffles the sub-section of the supplied slice starting from the first A or AAAA record and going
+// until the end. In other words, doesn't shuffle CNAME records at the start of the slice whose order
+// should be maintained.
 func shuffle(items *[]dns.RR) {
-
-	for i := range *items {
-		j := rand.Intn(i + 1)
+	min := 0
+	max := len(*items)
+	foundA := false
+	for i, record := range *items {
+		if !foundA && record.Header().Rrtype != dns.TypeA && record.Header().Rrtype != dns.TypeAAAA {
+			continue
+		}
+		foundA = true
+		if min == 0 {
```

Comment:
I think this makes the shuffle unbalanced and you want just `j := i+rand.Intn(max-i)` every time so each entry is swapped exactly once.